src/magneticField/turbulentField/PlaneWaveTurbulence.cpp

```diff
@@ -0,0 +1,403 @@
+		// *
+		// * argument reduction
+		// step 1: compute round(x), and store it in q
+		__m256d q = _mm256_round_pd(
+		    cos_arg, (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+
+		// we now compute s, which will be the input parameter to our polynomial
+		// approximation of cos(pi/2*x) between 0 and 1 the andnot_pd is just a
+		// fast way of taking the absolute value
+		__m256d s = _mm256_sub_pd(cos_arg, q);
+
+		// the following based on the int extraction process described here:
+		// https://stackoverflow.com/questions/41144668/how-to-efficiently-perform-double-int64-conversions-with-sse-avx/41223013
+		// we assume -2^51 <= q < 2^51 for this, which is unproblematic, as
+		// double precision has decayed far enough at that point that the cosine
+		// would be useless anyway.
+
+		// we now want to check whether q is even or odd, because the cosine is
+		// negative for odd qs, so we'll have to flip the final result. on an
+		// int, this is as simple as checking the 0th bit.
+		// => manipulate the double in such a way that we can do this.
+		// so, we add 2^52, such that the last digit of the mantissa is actually
+		// in the ones position. since q may be negative, we'll also add 2^51 to
+		// make sure it's positive. note that 2^51 is even and thus leaves
+		// evenness invariant, which is the only thing we care about here.
+
+		q = _mm256_add_pd(q, _mm256_set1_pd(0x0018000000000000));
+
+		// unfortunately, integer comparisons were only introduced in avx2, so
+		// we'll have to make do with a floating point comparison to check
+		// whether the last bit is set. however, masking out all but the last
+		// bit will result in a denormal float, which may either result in
+		// performance problems or just be rounded down to zero, neither of
+		// which is what we want here. To fix this, we'll mask in not only bit
+		// 0, but also the exponent (and sign, but that doesn't matter) of q.
+		// Luckily, the exponent of q is guaranteed to have the fixed value of
+		// 1075 (corresponding to 2^52) after our addition.
+
+		__m256d invert = _mm256_and_pd(
+		    q, _mm256_castsi256_pd(_mm256_set1_epi64x(0xfff0000000000001)));
+
+		// if we did have a one in bit 0, our result will be equal to 2^52 + 1
+		invert = _mm256_cmp_pd(
+		    invert, _mm256_castsi256_pd(_mm256_set1_epi64x(0x4330000000000001)),
+		    _CMP_EQ_OQ);
+
+		// finally, we need to turn invert and right_invert into masks for the
+		// sign bit on each final double, ie
+		invert = _mm256_and_pd(invert, _mm256_set1_pd(-0.0));
+
+		// TODO: clamp floats between 0 and 1? This would ensure that we never
+		// see inf's, but maybe we want that, so that things dont just fail
+		// silently...
+
+		// * end of argument reduction
+		// *******
+
+		// ******
+		// * evaluate the cosine using a polynomial approximation
+		// * the coefficients for this were generated using sleefs gencoef.c
+		// * I have no idea what I'm doing, so these coefficients are probably
```

Comment:
This should be rephrased before going into the master branch.

Reply:
Yes. Changed.